src/main/java/org/ehcache/integrations/shiro/EhcacheShiro.java

```diff
@@ -57,16 +57,41 @@ public V get(K k) throws CacheException {
   public V put(K k, V v) throws CacheException {
     trace("Putting object", k);
 
-    V previousValue = get(k);
-    cache.put(k, v);
+    V previousValue = null;
+
+    while (true) {
+      previousValue = get(k);
+      if (previousValue == null) {
+        if (cache.putIfAbsent(k, v) == null) {
+          break;
+        }
+      } else {
+        if (cache.replace(k, v) != null) {
+          break;
+        }
+      }
+    }
+
     return previousValue;
   }
 
   public V remove(K k) throws CacheException {
     trace("Removing object", k);
 
-    V previousValue = get(k);
-    cache.remove(k);
+    V previousValue = null;
+
+    while (true) {
+      previousValue = get(k);
+      if (previousValue == null) {
+        cache.remove(k);
```

Comment:
You can return directly here - null values are not supported so you can assume remove is not required.